src/krate.rs

```diff
@@ -81,6 +98,133 @@ pub struct CrateLinks {
     pub reverse_dependencies: String,
 }
 
+#[derive(Insertable, AsChangeset, Default)]
+#[table_name="crates"]
+pub struct NewCrate<'a> {
+    pub name: &'a str,
+    pub description: Option<&'a str>,
+    pub homepage: Option<&'a str>,
+    pub documentation: Option<&'a str>,
+    pub readme: Option<&'a str>,
+    pub repository: Option<&'a str>,
+    pub license: Option<&'a str>,
+    pub max_upload_size: Option<i32>,
+}
+
+impl<'a> NewCrate<'a> {
+    pub fn create_or_update(
```

Comment:
This is the same functionality as `find_or_insert`, but with a more clear name, and using Diesel. It's only used in tests currently.